sklearn/preprocessing/target.py

```diff
@@ -0,0 +1,220 @@
+        if self.check_inverse:
+            random_state = check_random_state(self.random_state)
+            n_subsample = min(10, y.shape[0])
+            subsample_idx = random_state.choice(range(y.shape[0]),
+                                                size=n_subsample,
+                                                replace=False)
+            if not np.allclose(
+                    y[subsample_idx],
+                    self.transformer_.inverse_transform(
+                        self.transformer_.transform(y[subsample_idx])),
+                    atol=1e-4):
+                raise ValueError("The provided functions or transformer are"
+                                 " not strictly inverse of each other. If"
+                                 " you are sure you want to proceed regardless"
+                                 ", set 'check_inverse=False'")
+
+    def fit(self, X, y, sample_weight=None):
+        """Fit the model according to the given training data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Training vector, where n_samples is the number of samples and
+            n_features is the number of features.
+
+        y : array-like, shape (n_samples,)
+            Target values.
+
+        sample_weight : array-like, shape (n_samples,) optional
+            Array of weights that are assigned to individual samples.
+            If not provided, then each sample is given unit weight.
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        y = check_array(y, ensure_2d=False)
+        self.y_ndim_ = y.ndim
+        if y.ndim == 1 and self.func is None:
+            y_2d = y.reshape(-1, 1)
+        else:
+            y_2d = y
+        self._fit_transformer(y_2d, sample_weight)
+        if self.regressor is None:
+            self.regressor_ = LinearRegression()
+        else:
+            self.regressor_ = clone(self.regressor)
+        if sample_weight is not None:
+            self.regressor_.fit(X, self.transformer_.fit_transform(y_2d),
+                                sample_weight=sample_weight)
+        else:
+            self.regressor_.fit(X, self.transformer_.fit_transform(y_2d))
+        return self
+
+    def predict(self, X):
+        """Predict using the base regressor, applying inverse.
+
+        The regressor is used to predict and the ``inverse_func`` or
+        ``inverse_transform`` is applied before returning the prediction.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            Samples.
+
+        Returns
+        -------
+        y_hat : array, shape = (n_samples,)
+            Predicted values.
+
+        """
+        check_is_fitted(self, "regressor_")
+        pred = self.transformer_.inverse_transform(self.regressor_.predict(X))
+        if self.y_ndim_ == 1 and self.func is None:
```

Comment:
To avoid the useless reshaping for `func` and `inverse_func` case. See [Joel's comment](https://github.com/scikit-learn/scikit-learn/pull/9041#discussion_r120775777)